common/src/comments.rs

```diff
@@ -2,91 +2,210 @@ use crate::{
     pos::Spanned,
     syntax_pos::{BytePos, Span},
 };
-use dashmap::{mapref::one::Ref, DashMap};
-use fxhash::FxBuildHasher;
+use fxhash::FxHashMap;
+use std::{cell::RefCell, rc::Rc, sync::Arc};
+
+/// Stores comment.
+///
+/// ## Implementation notes
+///
+/// Methods uses `(&self)` instead of `(&mut self)` for some reasons. Firstly,
+/// this is similar to the previous api. Secondly, typescript parser requires
+/// backtracking, which requires [Clone]. To avoid cloning large vectors, we
+/// must use [Rc<RefCell<Comments>>]. We have two option. We may implement it in
+/// the parser or in the implementation. If we decide to go with first option,
+/// we should pass [Comments] to parser, and as a result we need another method
+/// to take comments back. If we decide to go with second way, we can just pass
+/// [&Comments] to the parser. Thirdly, `(&self)` allows multi-threaded
+/// use-cases such as swc itself.
+///
+/// We use [Option] instead of no-op Comments implementation to avoid allocation
```

Comment:
AFAIK Empty vec doesn't allocate.

Reply:
You are correct, but to avoid allocation in parser, you need to determine if comment parser is enabled.

It can be done by passing comment map whilest passing a flag to parser, but Option<&dyn Comments> can be used as a comment map combined with a flag.